client/consensus/src/collators/lookahead.rs

```diff
@@ -74,8 +85,255 @@ use {
+    aux_data: OrchestratorAuraWorkerAuxData<P>,
+    inherent_providers: CIDP::InherentDataProviders,
+    keystore: &KeystorePtr,
+    orchestrator_client: Arc<OClient>,
+    orchestrator_tx_pool: Arc<FullPool<OBlock, OClient>>,
+    parent_header: <Block as BlockT>::Header,
+    orchestrator_slot_duration: SlotDuration,
+    container_slot_duration: SlotDuration,
+) -> Result<
+    <OBlock as BlockT>::Hash,
+    BuyCoreError<
+        <<OBlock as BlockT>::Header as HeaderT>::Number,
+        <FullPool<OBlock, OClient> as TransactionPool>::Error,
+    >,
+>
+where
+    Block: BlockT,
+    OBlock: BlockT,
+    P: Pair<Public = NimbusId> + Sync + Send + 'static,
+    P::Signature: TryFrom<Vec<u8>> + Member + Codec,
+    CIDP: CreateInherentDataProviders<Block, (PHash, PersistedValidationData)>
+        + Send
+        + 'static
+        + Clone,
+    CIDP::InherentDataProviders: Send + InherentDataProviderExt,
+    OClient: ProvideRuntimeApi<OBlock>
+        + HeaderMetadata<OBlock, Error = sp_blockchain::Error>
+        + HeaderBackend<OBlock>
+        + BlockBackend<OBlock>
+        + BlockIdTo<OBlock>
+        + 'static,
+    OClient::Api: TaggedTransactionQueue<OBlock>
+        + XCMCoreBuyerApi<OBlock, <<OBlock as BlockT>::Header as HeaderT>::Number, ParaId, NimbusId>,
+{
+    // We do nothing if this is not a parathread
+    if aux_data.slot_freq.is_none() {
+        return Err(BuyCoreError::NotAParathread);
+    }
+
+    let orchestrator_best_hash = orchestrator_client.info().best_hash;
+    let orchestrator_runtime_api = orchestrator_client.runtime_api();
+
+    let buy_core_slot_drift = orchestrator_client
+        .runtime_api()
+        .get_buy_core_slot_drift(orchestrator_best_hash)?;
+
+    // Convert drift in terms of container chain slots for parity between client side calculation and
+    // orchestrator runtime calculation
+    // We tried to made this calculation as generic as possible so that it can handle
+    // arbitrary slot timings as well and won't assume anything.
+    // Formula is: (Slot_drift_in_orchestrator_slot * orchestrator_slot_duration) / container_slot_duration
+    let buy_core_container_slot_drift = buy_core_slot_drift
+        .checked_mul(orchestrator_slot_duration.as_millis())
+        .and_then(|intermediate_result| {
+            intermediate_result.checked_div(container_slot_duration.as_millis())
+        })
+        .ok_or(BuyCoreError::SlotDriftConversionOverflow)?;
+
+    let current_container_slot = inherent_providers.slot();
+
+    let slot_claim = tanssi_claim_slot::<P, Block>(
+        aux_data,
+        &parent_header,
+        current_container_slot,
+        ClaimMode::ParathreadCoreBuying {
+            drift_permitted: buy_core_container_slot_drift.into(),
+        },
+        keystore,
+    )
+    .ok_or(BuyCoreError::UnableToClaimSlot)?;
+
+    let pubkey = slot_claim.author_pub;
+
+    orchestrator_runtime_api.is_core_buying_allowed(orchestrator_best_hash, para_id)??;
```

Comment:
I would pass the public key to `is_core_buying_allowed`. Just in case in the future we want to filter based of key